mgmtfn/k8splugin/driver.go

```diff
@@ -0,0 +1,371 @@
+}
+
+// epCleanUp deletes the ep from netplugin and netmaster
+func epCleanUp(req *epSpec) error {
+	// first delete from netplugin
+	// ignore any errors as this is best effort
+	netID := req.Network + "." + req.Tenant
+	err1 := netPlugin.DeleteEndpoint(netID + "-" + req.EndpointID)
+
+	// now delete from master
+	delReq := master.DeleteEndpointRequest{
+		TenantName:  req.Tenant,
+		NetworkName: req.Network,
+		ServiceName: req.Group,
+		EndpointID:  req.EndpointID,
+	}
+
+	var delResp master.DeleteEndpointResponse
+	err2 := cluster.MasterPostReq("/plugin/deleteEndpoint", &delReq, &delResp)
+
+	if err1 != nil {
+		return err1
+	}
+
+	return err2
+}
+
+// createEP creates the specified EP in contiv
+func createEP(req *epSpec) (*epAttr, error) {
+
+	// if the ep already exists, treat as error for now.
+	netID := req.Network + "." + req.Tenant
+	ep, err := netdGetEndpoint(netID + "-" + req.EndpointID)
+	if err == nil {
+		return nil, fmt.Errorf("EP %s already exists", req.EndpointID)
+	}
+
+	// Build endpoint request
+	mreq := master.CreateEndpointRequest{
+		TenantName:  req.Tenant,
+		NetworkName: req.Network,
+		ServiceName: req.Group,
+		EndpointID:  req.EndpointID,
+		ConfigEP: intent.ConfigEP{
+			Container:   req.EndpointID,
+			Host:        pluginHost,
+			ServiceName: req.Group,
+		},
+	}
+
+	var mresp master.CreateEndpointResponse
+	err = cluster.MasterPostReq("/plugin/createEndpoint", &mreq, &mresp)
+	if err != nil {
+		epCleanUp(req)
+		return nil, err
+	}
+
+	log.Infof("Got endpoint create resp from master: %+v", mresp)
+
+	// Ask netplugin to create the endpoint
+	err = netPlugin.CreateEndpoint(netID + "-" + req.EndpointID)
+	if err != nil {
+		log.Errorf("Endpoint creation failed. Error: %s", err)
+		epCleanUp(req)
+		return nil, err
+	}
+
+	ep, err = netdGetEndpoint(netID + "-" + req.EndpointID)
+	if err != nil {
+		epCleanUp(req)
+		return nil, err
+	}
+
+	log.Debug(ep)
+	// need to get the subnetlen from nw state.
+	nw, err := netdGetNetwork(netID)
+	if err != nil {
+		return nil, err
+	}
+
+	epResponse := epAttr{}
+	epResponse.PortName = ep.PortName
+	epResponse.IPAddress = ep.IPAddress + "/" + strconv.Itoa(int(nw.SubnetLen))
+
+	return &epResponse, nil
+}
+
+// getLink is a wrapper that fetches the netlink corresponding to the ifname
+func getLink(ifname string) (netlink.Link, error) {
+	// find the link
+	link, err := netlink.LinkByName(ifname)
+	if err != nil {
+		if !strings.Contains(err.Error(), "Link not found") {
+			log.Errorf("unable to find link %q. Error: %q", ifname, err)
+			return link, err
+		}
+		// try once more as sometimes (somehow) link creation is taking
+		// sometime, causing link not found error
+		time.Sleep(1 * time.Second)
+		link, err = netlink.LinkByName(ifname)
+		if err != nil {
+			log.Errorf("unable to find link %q. Error %q", ifname, err)
+		}
+		return link, err
+	}
+	return link, err
+}
+
+// nsToPID is a utility that extracts the PID from the netns
+func nsToPID(ns string) (int, error) {
```

Comment:
As we don't know the exact needs of mesos plugin at this point, I will leave this to @unclejack to refactor and move as appropriate.